keras_nlp/samplers/sampler.py

```diff
@@ -344,15 +380,16 @@ def one_step(current_index, prompt, mask):
             )
 
             current_index = tf.add(current_index, 1)
-            return (current_index, prompt, mask)
+            return (current_index, prompt, mask, unfinished)
 
         # Run a while loop till `max_length` of tokens has been generated.
-        current_index, prompt, mask = tf.while_loop(
-            cond=lambda current_index, prompt, mask: tf.less(
+        current_index, prompt, mask, unfinished = tf.while_loop(
+            cond=lambda current_index, prompt, mask, unfinished: tf.less(
```

Comment:
What if we did this all in the cond, use `original_padding_mask`, and don't add any loop vars. Something like this.

```python
new_end_tokens = (prompt == end_token_id) & (original_padding_mask == False)
sequence_done = tf.reduce_any(new_end_tokens, axis=-1)
all_done = tf.reduce_all(sequence_done)
```

Haven't actually run that code yet, but something like that feels like it would be a lot cleaner. And the overall cost of doing a sweep of the generated prompt should be peanuts compared to the forward pass, so I don't think we need to worry about performance too much here.

Reply:
sg! changed.